Use the examples directory for example plugin code

ptgott opened the pull request:
Closes #27696

Also edit the Access Request plugin API guide to use this directory, rather than having the reader copy/paste individual code snippets. This makes the guide easier to follow, and users will have a compilable example before they proceed through the guide.

See [this discussion](https://github.com/gravitational/teleport/pull/23204#pullrequestreview-1344681896) for context. I will also apply similar changes to my in-progress API guides.

ptgott committed:
Use the examples directory for example plugin code

Also edit the Access Request plugin API guide to use this directory,
rather than having the reader copy/paste individual code snippets. This
makes the guide easier to follow, and users will have a compilable
example before they proceed through the guide.

alexfornuto reviewed (approved):
I'm not a go programmer so I don't know if this makes sense, but if the sections that we break out and describe individually in these guides could be separate files imported into a main file, we could use partials inside the code blocks and ensure that the docs remain up-to-date if/when the examples change.

ptgott commented:
@alexfornuto I think that's a really good idea, though it will result in a slightly strange-looking program, since we'll have a collection of very small `.go` files. What do you think, @zmb3 ?

ptgott committed:
Run make fix-license

ptgott committed:
Run make fix-imports

ptgott committed:
Fix spelling

ptgott committed:
Run go mod tidy

Joerger commented on examples/access-plugin-minimal/go.mod:
```diff
@@ -0,0 +1,64 @@
+	google.golang.org/grpc v1.53.0
+)
+
+require (
```
nit: For a "minimal" example, this is a lot more imports than I would expect. I'm guessing a lot of these are coming from `github.com/gravitational/teleport-plugins`? Would it be worthwhile to remove that dependency and just use `github.com/gravitational/teleport/api` directly, in a similar fashion to the workflows example?

ptgott replied:
The `teleport-plugins` imports are:

```
	"github.com/gravitational/teleport-plugins/lib"
	"github.com/gravitational/teleport-plugins/lib/watcherjob"
```

I like the simplicity of the `watcherjob.NewJob` interface, and it looks like we need `lib.NewProcess(ctx)` to run it. 

I know we're also migrating these libraries to `gravitational/teleport`, though, so I'm guessing we'll need to change these imports at some point anyway (as of now, this code compiles successfully for me on a fresh `ubuntu:20.04` container after installing git and go and checking out this branch—not sure when that will stop working).

What would we need to do to reduce the number of imports?

Joerger replied:
> I like the simplicity of the watcherjob.NewJob interface, and it looks like we need lib.NewProcess(ctx) to run it.

Fair enough. Ideally you'd be able to use the `WatcherJob` interface without importing all the other junk, but that's not really your problem.

> I know we're also migrating these libraries to gravitational/teleport, though, so I'm guessing we'll need to change these imports at some point anyway

Actually, it looks like this was already moved into `gravitational/teleport/integrations`. You can probably easily switch the import, but the current code won't stop working as the old teleport-plugin go module will be available indefinitely.

Unfortunately, that means you will be importing even more code that you don't need, as we are back to the first problem listed here - https://github.com/gravitational/teleport/blob/master/rfd/0010-api.md#why.

Joerger commented on examples/access-plugin-minimal/main.go:
```diff
@@ -0,0 +1,274 @@
+		panic(err)
+	}
+
+	gs := googleSheetsPlugin{
```
nit: From the perspective of using this example as a base for one's own custom plugin, it would be cool to make the code a bit more modular. For example, rather than having a `googleSheetsPlugin`, we could have something like:

```
type accessRequestPlugin struct {
	teleportClient *client.Client
        eventHandler interface {
                HandleEvent(ctx context.Context, event types.Event) error
        }
}

type googleSheetsClient struct {
        sheetsClient   *sheets.SpreadsheetsService
}

func (*googleSheetsClient) HandleEvent(ctx context.Context, event types.Event) error {
       ...
}

plugin := accessRequestPlugin{
       teleportClient: ...,
       eventHandler: googleSheetsClient{
              sheetsClient: ...,
       }
}
```

Then the user could just swap out `eventHandler` with their own implementation very intuitively. This would also make splitting the code into separate files more natural.

I think this PR is a big improvement regardless, so this is more of a follow up suggestion than a blocker.

ptgott replied:
This sounds great, especially if we want to make it the recommended way for a user to build/test their own Access Request plugin. If this is the way we want to go, I can look into making this change. 

When writing this guide, though, I had imagined the example as something that would demonstrate some key libraries in a runnable way, but not the basis for an actual production plugin. Is there anything else we'd need to add to the `accessRequestPlugin` type to make it more generally usable?

Joerger replied:
Nothing off the top of my head, just having a teleport Client and some customizable event handler seems sufficient as a base.

Joerger reviewed (commented):
This is an awesome guide and sample project!

> I'm not a go programmer so I don't know if this makes sense, but if the sections that we break out and describe individually in these guides could be separate files imported into a main file, we could use partials inside the code blocks and ensure that the docs remain up-to-date if/when the examples change.

+1 to this idea. I wouldn't split it into 10 files or anything, but I think you could reasonbly chunk it like this or similar:
 - `func main`
 - `run` and `handleEvent`
 - `updateSpreadsheet`
 - `updateRow` and `createRow`
 - `makeRowData`

> ... another aspect of this is that we would need to include `package main` at the top of each partial, which will also look strange within the guide itself.

IMO overall it's a worthwhile tradeoff, and as a Go programmer I wouldn't be confused by it.

ptgott committed:
Extract Access Request plugin example to partials

This way, we can reuse the actual program in the Access Request plugin
API guide and avoid unintended discrepancies and drift.

ptgott commented:
Going to put this one in draft and plan to rework the example application so it doesn't rely on `watcherjob`. Instead, we can use `teleport/api/types.Events`.

ptgott committed:
Use types.Events.NewWatcher instead of watcherjob

Need to test this out, but it compiles

ptgott committed:
Remove outdated information

- Types that are no longer reachable via public interfaces
- The description of the demo implementation that used the old
  `watcherjob` package

ptgott committed:
Update text to reflect new `run` logic

ptgott committed:
Make the example program more modular

Respond to Joerger feedback

alexfornuto reviewed (approved):
Copy changes look good, with one nit.

ptgott committed:
Respond to alexfornuto feedback

Joerger commented on examples/access-plugin-minimal/watcherjob.go:
```diff
@@ -0,0 +1,65 @@
+		},
+	})
+
+	if err != nil {
+		return trace.Wrap(err)
+	}
```
nit:
```suggestion
	if err != nil {
		return trace.Wrap(err)
	}
	defer watch.Close()
```

Joerger commented on examples/access-plugin-minimal/watcherjob.go:
```diff
@@ -0,0 +1,65 @@
+	ctx := context.Background()
+
+	watch, err := p.TeleportClient.NewWatcher(ctx, types.Watch{
+		Name: "Access Requests",
```
nit: The name is just ignored by the client currently so you can omit it. The Auth server sets the name of the watcher itself internally, just for debugging purposes.
```suggestion
```

Joerger commented on examples/access-plugin-minimal/watcherjob.go:
```diff
@@ -0,0 +1,65 @@
+	for {
+		select {
+		case e := <-watch.Events():
+			p.EventHandler.HandleEvent(ctx, e)
```
nit: handle error
```suggestion
			if err := p.EventHandler.HandleEvent(ctx, e); err != nil {
				return trace.Wrap(err)
			}
```

Joerger commented on examples/access-plugin-minimal/main.go:
```diff
@@ -0,0 +1,62 @@
+			grpc.WithReturnConnectionError(),
+		},
+	})
+	if err != nil {
+		panic(err)
+	}
```
nit:
```suggestion
	if err != nil {
		panic(err)
	}
	defer teleport.Close()
```

Joerger commented on examples/access-plugin-minimal/main.go:
```diff
@@ -0,0 +1,62 @@
+		panic(err)
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, initTimeout)
+	defer cancel()
+
```
nit: I don't think this is needed, right?
```suggestion
```

Joerger commented on docs/pages/api/access-plugin.mdx:
```diff
@@ -237,37 +209,19 @@ The plugin imports the following third-party code:
-	sheetsClient   *sheets.SpreadsheetsService
-	teleportClient *client.Client
-}
+(!examples/access-plugin-minimal/types.go!)
```
nit: Is it possible to choose a specific line range to show so we can cut out the copyright comment at the top? Doesn't seem too awful to keep it, either.

ptgott replied:
Sadly, no—the docs engine includes the entire partial

Joerger reviewed (approved):
LGTM other than some nit picks

ptgott committed:
Apply suggestions from code review

Co-authored-by: Brian Joerger <bjoerger@goteleport.com>

zmb3 commented on examples/access-plugin-minimal/types.go:
```diff
@@ -0,0 +1,49 @@
+	}
+}
+
+type googleSheetsClient struct {
```
I would move this type and its methods into a file called `sheets.go`. It's generally preferred to organize code in a way that code related to a particular function is grouped, rather than having things like `types.go`, `consts.go`, `methods.go`, etc.

zmb3 commented on examples/access-plugin-minimal/createrow.go:
```diff
@@ -0,0 +1,59 @@
+	if resp.HTTPStatusCode == 201 || resp.HTTPStatusCode == 200 {
+		fmt.Println("Successfully created a row")
+	} else {
+		fmt.Printf(
```
If you didn't get a successful response, should you be returning an error here? 

For example, if your credentials aren't valid and you get a 403 forbidden, you didn't create the row successfully, so returning nil seems odd.

If you make the change, I'd also suggest moving the printlns out of here and let the caller print based on the returned value.

zmb3 commented on examples/access-plugin-minimal/makerowdata.go:
```diff
@@ -0,0 +1,76 @@
+func stringPtr(s string) *string { return &s }
+
+func (g *googleSheetsClient) makeRowData(ar types.AccessRequest) *sheets.RowData {
+	requestState, ok := requestStates[ar.GetState()]
+
+	// Could not find a state, but this is still a valid Access Request
+	if !ok {
+		requestState = requestStates[types.RequestState_NONE]
+	}
```
```suggestion
	requestState := requestStates[ar.GetState()]
```

`types.RequestState_NONE` is going to be the default value, so you can simplify this.

zmb3 commented on examples/access-plugin-minimal/types.go:
```diff
@@ -0,0 +1,49 @@
+)
+
+const (
+	proxyAddr     string = ""
```
At first I was thinking "why in the world is there a const for an empty string?"

How about some type of placeholder (or even comment) to make it clear that this needs to be updated?

```suggestion
	proxyAddr     string = "CHANGE ME"
```

zmb3 commented on examples/access-plugin-minimal/updaterow.go:
```diff
@@ -0,0 +1,62 @@
+	if resp.HTTPStatusCode == 201 || resp.HTTPStatusCode == 200 {
+		fmt.Println("Successfully updated a row")
+	} else {
+		fmt.Printf(
```
Same comment here as for create row.

zmb3 commented on examples/access-plugin-minimal/updatespreadsheet.go:
```diff
@@ -0,0 +1,52 @@
+				if err := g.updateRow(ar, int64(i)); err != nil {
+					return trace.Wrap(err)
+				}
+				fmt.Println("Updated a spreadsheet row.")
```
On success, you're going to end up printing this message twice, since `updateRow` also prints out a message when it successfully updates a row.

ptgott committed:
Respond to zmb3 feedback

- Split up "types.go". Reserve a single file for configuration values so
  these are visible in a single place within the guide.
- Return an error on an unsuccessful HTTP request when creating or
  updating a row
- Simplify requestStates lookup
- Clearly mark values that a user must change
- Update the text of the guide to match changes to the program

ptgott committed:
Spell fixes

ptgott commented:
@zmb3 Friendly ping to see if I resolved your feedback

zmb3 commented on examples/access-plugin-minimal/createrow.go:
```diff
@@ -0,0 +1,56 @@
+		return nil
+	}
+
+	return trace.Wrap(
+		fmt.Errorf("Unexpected response code creating a row: %v\n",
+			resp.HTTPStatusCode),
+	)
```
```suggestion
	return trace.Errorf("Unexpected response code creating a row: %v", resp.HTTPStatusCode)
```

- definitely don't want the `\n` in the error message
- `trace.Errorf` should be sufficient

zmb3 commented on examples/access-plugin-minimal/watcherjob.go:
```diff
@@ -0,0 +1,77 @@
+	if r.GetState() == types.RequestState_PENDING {
+		if err := g.createRow(r); err != nil {
+			return err
+		} else {
```
You can eliminate the `else` here. The fact that the if case ends with a return means everything after that is a de-facto else.

zmb3 commented on examples/access-plugin-minimal/watcherjob.go:
```diff
@@ -0,0 +1,77 @@
+
+	if err := g.updateSpreadsheet(r); err != nil {
+		return err
+	} else {
```
Eliminate the else.

ptgott committed:
Respond to zmb3 feedback
